Implement Base Node Wallet RPC service

philipr-za opened the pull request:
## Description

Implement a Wallet RPC service in the Base Node

This PR implements an RPC service in the base node for the wallet to use when submitting and querying transactions. The RPC service provides two calls: 
`submit_transaction(tx)`: Will be used by a wallet to submit a transaction to the Base Node and the response informs the wallet if the tx was accepted into the mempool or it was rejected and why
`transaction_query(excess_sig)`: Lets the wallet request the status of a transaction via its excess_sig. The response will tell the wallet whether the transaction is Not Stored, in the Unconfirmed Pool or if it is Mined it will also return the block header and number of confirmations.

## Motivation

This RPC service represents the new interface that a wallet will use to interact with base nodes while trying to broadcast a transaction. This will replace the current fire-and-forget type messaging provided by the p2p comms layer which makes the logic for receiving the response to the query tricky.

## How Has This Been Tested?
The `test_base_node_wallet_rpc()` is provided that tests both the RPC calls using a test blockchain and mempool back end.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

delta1 commented:
This looks great, but `build wallet` failed in circle ci and it seems there are some test compilation failures too

philipr-za committed:
Implement a Wallet RPC service in the Base Node

This PR implements an RPC service in the base node for the wallet to use when submitting and querying transactions. The RPC service provides two calls: 
`submit_transaction(tx)`: Will be used by a wallet to submit a transaction to the Base Node and the response informs the wallet if the tx was accepted into the mempool or it was rejected and why
`transaction_query(excess_sig)`: Lets the wallet request the status of a transaction via its excess_sig. The response will tell the wallet whether the transaction is Not Stored, in the Unconfirmed Pool or if it is Mined it will also return the block header and number of confirmations.

This RPC service represents the new interface that a wallet will use to interact with base nodes while trying to broadcast a transaction. This will replace the current fire-and-forget type messaging provided by the p2p comms layer which makes the logic for receiving the response to the query tricky.

The `test_base_node_wallet_rpc()` is provided that tests both the RPC calls using a test blockchain and mempool back end.

philipr-za commented:
Urgh, those feature flags in core are a pain in the ass.

philipr-za commented:
woops

stringhandler commented on base_layer/core/src/base_node/rpc/service.rs:
```diff
@@ -0,0 +1,194 @@
+                confirmations: 0,
+            },
+            TxStorageResponse::ReorgPool |
+            TxStorageResponse::NotStoredOrphan |
```
Is it possible for the mempool to ever return these? Surely the mempool will not know the reason something is not stored, because it has not stored it

philipr-za replied:
The validator tells it what the issue was, that's how it used to decide which pool to put it in.

philipr-za replied:
oh on the query you are right its not possible

philipr-za replied:
I agree on this change and I got started but turns out that the same struct is used for the service handle interface that is used by the proto of the mempool service so making an update to way the handle returns involves updating everything in the repo that touches the proto too which is extensive.  Should I do it in this PR?

stringhandler reviewed (approved):
Looks good. I'd say that TxStorageResponse should not be reused for querying, and only used for submitting.
